Early rough draft of the benchit user guide (#1)
groq/mlagility: docs/user_guide.md

# MLABench User Guide

MLAgility Benchmarking (MLABench) is a Python package that provides a CLI, `benchit`, and API, `benchit()` for benchmarking machine learning and deep learning models. This document reviews the functionality provided by MLABench.

MLABench currently supports the following combinations of runtimes and devices:

<span id="devices-runtimes-table">

| Device Type | Device arg | Runtime      | Runtime arg | Specific Devices                                      |
|-------------|------------|--------------|-------------|-------------------------------------------------------|

> | Nvidia GPU  | nvidia     | TensorRT     | trt         | Any Nvidia GPU supported by TensorRT 8.5.2            |

Reviewer: ```suggestion
| Nvidia GPU  | nvidia     | TensorRT     | trt         | Any Nvidia GPU supported by TensorRT >= 8.5.2            |
```

> | x86 CPU     | x86        | ONNX Runtime | ort         | Any Intel or AMD CPU supported by ONNX Runtime 1.14.0 |

Reviewer: ```suggestion
| x86 CPU     | x86        | ONNX Runtime | ort         | Any Intel or AMD CPU supported by ONNX Runtime >= 1.13.1 |
```

For example, the following script:

```python
from mlabench import benchit

model = YourModel()
results = model(**inputs)
perf = benchit(model, inputs)
```

Will print an output like this:

```
> Performance of YourModel on device Intel® Xeon® Platinum 8380 is:
> latency: 0.033 ms
> throughput: 21784.8 ips
```

`benchit()` also returns a `Performance` object that includes members:
 - `latency_units`: unit of time used for measuring latency, which is set to `milliseconds (ms)`.
 - `latency`: average benchmarking latency, measured in `latency_units`.
 - `throughput_units`: unit used for measuring throughput, which is set to `inferences per second (IPS)`.
 - `throughput`: average benchmarking throughput, measured in `throughput_units`.

 _Note_: The benchmarking methodology is defined [here](#benchmark).

## Definitions

MLABench uses the following definitions throughout.

### Model

A **model** is a PyTorch (torch.nn.Module) or Keras (tf.keras.Model) object that has been instantiated in a Python script.

- Examples: BERT-Base, ResNet-50, etc.

### Device

A **device** is a piece of hardware capable of running a model.

> - Examples: Nvidia A100 40GB, Intel Xeon Platinum 8380, Groq GroqChip1

Reviewer: ```suggestion
- Examples: Nvidia A100, Intel Xeon Platinum 8380, Groq GroqChip1
```

Author: Nvidia A100 40GB is a specific device, while Nvidia A100 is a family of devices. Performance is different between A100 40GB and 80GB because they have different memory bandwidth. Wont fix.

### Runtime

A **runtime** is a piece of software that executes a model on a device.

- Different runtimes can produce different performance on the same device because A) runtimes often optimize the model prior to execution, and B) the runtime is responsible for orchestrating data movement, device invocation, etc.
- Examples: ONNX Runtime, TensorRT, PyTorch Eager Execution, etc.

### Analysis

**Analysis** is the process by which `benchit` inspects a Python script and identifies the PyTorch/Keras models within.

> `benchit` performs analysis by running your script and profiling each line of code that is called. When a model object (see [Model](#model) is encountered, the `benchit` CLI can inspect it to gather statistics (such as the number of parameters in the model) and/or pass it to the `benchit()` API for benchmarking.

Reviewer: ```suggestion
`benchit` performs analysis by running your script and profiling each line of code. When a model object (see [Model](#model) is encountered, the `benchit` CLI can inspect it to gather statistics (such as the number of parameters in the model) and/or pass it to the `benchit()` API for benchmarking.
```

By default, `benchit()` will run the model 100 times to collect the following statistics:
1. Mean Latency, in milliseconds (ms): the average time it takes the runtime/device combination to execute the model/inputs combination once. This includes the time spent invoking the device and transferring the model's inputs and outputs between host memory and the device (when applicable).
1. Throughput, in inferences per second (IPS):  the number of times the model/inputs combination can be executed on the runtime/device combination per second.
    - __Note__: `benchit()` is not aware of whether `inputs` is a single input or a batch of inputs. If your `inputs` is actually a batch of inputs, you should multiply `benchit()`'s reported IPS by the batch size.

## Devices and Runtimes

MLABench can be used to benchmark a model across a variety of runtimes and devices, as long as the device is available and the device/runtime combination is supported by MLABench.

### Available Devices

MLABench supports benchmarking on both locally installed devices, as well as devices on remote machines (e.g., cloud VMs).

If you are using a remote machine, it must:

> - be running

Reviewer: Replace `be running` with Turned ON?
